web/web.go

```diff
@@ -0,0 +1,773 @@
+			}
+			delete(sp.ColDefs, sPk.Col)
+			for i, pk := range sp.Pks {
+				if pk.Col == sPk.Col {
+					sp.Pks = removePk(sp.Pks, i)
+					break
+				}
+			}
+		}
+		sp.Pks = append(sp.Pks, ddl.IndexKey{Col: colName, Desc: false})
+	}
+	app.conv.SpSchema[table] = sp
+}
+
+func updateType(newType, table, newColName, srcTableName string, w http.ResponseWriter) {
+	sp := app.conv.SpSchema[table]
+	srcColName := app.conv.ToSource[table].Cols[newColName]
+	srcCol := app.conv.SrcSchema[srcTableName].ColDefs[srcColName]
+	var ty ddl.Type
+	var issues []internal.SchemaIssue
+	switch app.driver {
+	case "mysql", "mysqldump":
+		ty, issues = toSpannerTypeMySQL(srcCol.Type.Name, newType, srcCol.Type.Mods)
+	case "pg_dump", "postgres":
+		ty, issues = toSpannerTypePostgres(srcCol.Type.Name, newType, srcCol.Type.Mods)
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+		return
+	}
+	if len(srcCol.Type.ArrayBounds) > 1 {
+		ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+		issues = append(issues, internal.MultiDimensionalArray)
+	}
+	if srcCol.Ignored.Default {
+		issues = append(issues, internal.DefaultValue)
+	}
+	if srcCol.Ignored.AutoIncrement {
+		issues = append(issues, internal.AutoIncrement)
+	}
+	if len(issues) > 0 {
+		app.conv.Issues[srcTableName][srcCol.Name] = issues
+	}
+	ty.IsArray = len(srcCol.Type.ArrayBounds) == 1
+	colDef := sp.ColDefs[newColName]
+	colDef.T = ty
+	sp.ColDefs[newColName] = colDef
+	app.conv.SpSchema[table] = sp
+}
+
+func updateNotNull(notNullChange, table, newColName string) {
+	sp := app.conv.SpSchema[table]
+	switch notNullChange {
+	case "ADDED":
+		spColDef := sp.ColDefs[newColName]
+		spColDef.NotNull = true
+		sp.ColDefs[newColName] = spColDef
+	case "REMOVED":
+		spColDef := sp.ColDefs[newColName]
+		spColDef.NotNull = false
+		sp.ColDefs[newColName] = spColDef
+	default:
+		// We skip this.
+	}
+	app.conv.SpSchema[table] = sp
+}
+
+func rateSchema(cols, warnings int64, missingPKey bool) string {
+	good := func(total, badCount int64) bool { return badCount < total/20 }
+	ok := func(total, badCount int64) bool { return badCount < total/3 }
+	switch {
+	case cols == 0:
+		return "GRAY"
+	case warnings == 0 && !missingPKey:
+		return "GREEN"
+	case warnings == 0 && missingPKey:
+		return "BLUE"
+	case good(cols, warnings) && !missingPKey:
+		return "BLUE"
+	case good(cols, warnings) && missingPKey:
+		return "BLUE"
+	case ok(cols, warnings) && !missingPKey:
+		return "YELLOW"
+	case ok(cols, warnings) && missingPKey:
+		return "YELLOW"
+	case !missingPKey:
+		return "ORANGE"
+	default:
+		return "ORANGE"
+	}
+}
+
+func getFilePrefix(now time.Time) (string, error) {
+	dbName := app.dbName
+	var err error
+	if dbName == "" {
+		dbName, err = conversion.GetDatabaseName(app.driver, now)
+		if err != nil {
+			return "", fmt.Errorf("Can not create database name : %v", err)
+		}
+	}
+	filePrefix := dbName + "."
+	return filePrefix, nil
+}
+
+type App struct {
```

Comment:
Would SessionState be a better name that App?
And similarly for the var a few lines below: would it be clearer as sessionState?

Reply:
agreed, changed accordingly